In [60]:
import pandas as pd 
import numpy as np
import matplotlib.pyplot as pld
import seaborn as sns
import sqlite3

# Exercise 1

Using the same mtcars data, please execute the following:

**1.1** Select unique values in the year column and order them in acsending order. How many unique values are there?

**1.2** Create a column with vehicle age values

**1.3** If a vehicle is older than 45 years, then label it as "45+". Otherwise, label it as "<45". Create a new column that takes these two values

**1.4** How many vehicles are in our dataset that are under 45 and 45 and over?

**1.5** List all unique vehicle name values that have 8 cylinders with horsepower that is greater than 125 but doesn't exceed 200


In [61]:
# Establish a connection to the 'mtcars.sqlite' database

connection = sqlite3.connect('mtcars.sqlite')

In [62]:
# Execute the query to get the names of all tables in the database

pd.read_sql("SELECT name FROM sqlite_master WHERE type = 'table'",connection)

name
0       results
1        auto_2
2        test_2
3       song_db
4   test_output
5  test_output2
6   songs_table

In [63]:
# Select all rows and columns from the 'results' table

pd.read_sql_query('SELECT * FROM results',connection)

mpg  cylinders  displacement horsepower  weight  acceleration  year  \
0    18.0          8         307.0      130.0  3504.0          12.0    70   
1    15.0          8         350.0      165.0  3693.0          11.5    70   
2    18.0          8         318.0      150.0  3436.0          11.0    70   
3    16.0          8         304.0      150.0  3433.0          12.0    70   
4    17.0          8         302.0      140.0  3449.0          10.5    70   
..    ...        ...           ...        ...     ...           ...   ...   
392  27.0          4         140.0       86.0  2790.0          15.6    82   
393  44.0          4          97.0       52.0  2130.0          24.6    82   
394  32.0          4         135.0       84.0  2295.0          11.6    82   
395  28.0          4         120.0       79.0  2625.0          18.6    82   
396  31.0          4         119.0       82.0  2720.0          19.4    82   

     origin                       name  
0         1  chevrolet chevelle malibu  
1         1          buick skylark 320  
2         1         plymouth satellite  
3         1              amc rebel sst  
4         1                ford torino  
..      ...                        ...  
392       1            ford mustang gl  
393       2                  vw pickup  
394       1              dodge rampage  
395       1                ford ranger  
396       1                 chevy s-10  

[397 rows x 9 columns]

In [64]:
# 1.1 - Unique values in the year column in acsending order

pd.read_sql_query(
                    """
                        SELECT DISTINCT year -- Select unique years
                        FROM results
                        ORDER BY year ASC -- Order the years in ascending order
                    """,connection)

year
0     70
1     71
2     72
3     73
4     74
5     75
6     76
7     77
8     78
9     79
10    80
11    81
12    82

In [65]:
# 1.1 - How many unique values are there?

pd.read_sql_query(
                    """
                        SELECT COUNT(DISTINCT year) AS Numbre_of_unique_values -- Count the number of unique years
                        FROM results
                    """,connection)

Numbre_of_unique_values
0                       13

In [66]:
# 1.2 - Create a column with vehicle age values

pd.read_sql_query(
                    """
                        SELECT *,                           -- Select all columns
                        (2024-(1900+year)) AS vehicle_age   -- Calculate the age of the vehicle
                        FROM results
                    """,connection)

mpg  cylinders  displacement horsepower  weight  acceleration  year  \
0    18.0          8         307.0      130.0  3504.0          12.0    70   
1    15.0          8         350.0      165.0  3693.0          11.5    70   
2    18.0          8         318.0      150.0  3436.0          11.0    70   
3    16.0          8         304.0      150.0  3433.0          12.0    70   
4    17.0          8         302.0      140.0  3449.0          10.5    70   
..    ...        ...           ...        ...     ...           ...   ...   
392  27.0          4         140.0       86.0  2790.0          15.6    82   
393  44.0          4          97.0       52.0  2130.0          24.6    82   
394  32.0          4         135.0       84.0  2295.0          11.6    82   
395  28.0          4         120.0       79.0  2625.0          18.6    82   
396  31.0          4         119.0       82.0  2720.0          19.4    82   

     origin                       name  vehicle_age  
0         1  chevrolet chevelle malibu           54  
1         1          buick skylark 320           54  
2         1         plymouth satellite           54  
3         1              amc rebel sst           54  
4         1                ford torino           54  
..      ...                        ...          ...  
392       1            ford mustang gl           42  
393       2                  vw pickup           42  
394       1              dodge rampage           42  
395       1                ford ranger           42  
396       1                 chevy s-10           42  

[397 rows x 10 columns]

In [67]:
# 1.3 - Column that takes these two values ("45+" and "<45")

pd.read_sql_query(
                    """
                        SELECT *,
                                CASE
                                    WHEN vehicle_age >= 45 THEN '45+'        -- Assign '45+' if vehicle_age is 45 or more
                                    ELSE '<45'                               -- Otherwise, assign '<45'
                                END Categorical_age                          -- Alias the result as 'Categorical_age'
                        FROM 
                            (
                                SELECT *,(2024-(1900+year)) AS vehicle_age   -- Calculate the vehicle age
                                FROM results 
                             )
                    """,connection)

mpg  cylinders  displacement horsepower  weight  acceleration  year  \
0    18.0          8         307.0      130.0  3504.0          12.0    70   
1    15.0          8         350.0      165.0  3693.0          11.5    70   
2    18.0          8         318.0      150.0  3436.0          11.0    70   
3    16.0          8         304.0      150.0  3433.0          12.0    70   
4    17.0          8         302.0      140.0  3449.0          10.5    70   
..    ...        ...           ...        ...     ...           ...   ...   
392  27.0          4         140.0       86.0  2790.0          15.6    82   
393  44.0          4          97.0       52.0  2130.0          24.6    82   
394  32.0          4         135.0       84.0  2295.0          11.6    82   
395  28.0          4         120.0       79.0  2625.0          18.6    82   
396  31.0          4         119.0       82.0  2720.0          19.4    82   

     origin                       name  vehicle_age Categorical_age  
0         1  chevrolet chevelle malibu           54             45+  
1         1          buick skylark 320           54             45+  
2         1         plymouth satellite           54             45+  
3         1              amc rebel sst           54             45+  
4         1                ford torino           54             45+  
..      ...                        ...          ...             ...  
392       1            ford mustang gl           42             <45  
393       2                  vw pickup           42             <45  
394       1              dodge rampage           42             <45  
395       1                ford ranger           42             <45  
396       1                 chevy s-10           42             <45  

[397 rows x 11 columns]

In [68]:
# 1.4 - Vehicle older than 45 years, with label "45+"

pd.read_sql_query(
                    """
                        SELECT name,vehicle_age                            -- Select columns 'name' and 'vehicle_age'
                        FROM                                               -- Subquery to calculate vehicle age
                            (                                               
                                SELECT *,(2024-(1900+year)) AS vehicle_age -- Calculate vehicle age
                                FROM results  
                            )
                        WHERE vehicle_age >= 45                            -- Filter for vehicles older than or equal to 45
                    """,connection)

name  vehicle_age
0    chevrolet chevelle malibu           54
1            buick skylark 320           54
2           plymouth satellite           54
3                amc rebel sst           54
4                  ford torino           54
..                         ...          ...
304         fiat strada custom           45
305      buick skylark limited           45
306         chevrolet citation           45
307  oldsmobile omega brougham           45
308            pontiac phoenix           45

[309 rows x 2 columns]

In [69]:
# 1.4 - Count of Vehicle older than 45 years, with label "45+" (METHOD 1)

pd.read_sql_query(
                  """
                     SELECT COUNT(*) AS Count_of_Vehicles_older_than_45 -- Count the number of vehicles older than 45 years
                     FROM 
                        (
                            SELECT *,(2024-(1900+year)) AS vehicle_age  -- Calculate vehicle age
                            FROM results  
                        )
                     WHERE vehicle_age >= 45                            -- Filter for vehicles older than or equal to 45
                 """,connection)

Count_of_Vehicles_older_than_45
0                              309

In [70]:
# 1.4 - Vehicle youngest than 45 years, with label "<45"
pd.read_sql_query(
                    """
                        SELECT name,vehicle_age                            -- Select columns 'name' and 'vehicle_age'
                        FROM                                               -- Subquery to calculate vehicle age
                            (
                                SELECT *,(2024-(1900+year)) AS vehicle_age -- Calculate vehicle age
                                FROM results  
                            )
                        WHERE vehicle_age < 45                             -- Filter for vehicles younger than 45
                    """,connection)

name  vehicle_age
0               vw rabbit           44
1   toyota corolla tercel           44
2      chevrolet chevette           44
3              datsun 310           44
4      chevrolet citation           44
..                    ...          ...
83        ford mustang gl           42
84              vw pickup           42
85          dodge rampage           42
86            ford ranger           42
87             chevy s-10           42

[88 rows x 2 columns]

In [71]:
# 1.4 - Count of Vehicle youngest than 45 years, with label "<45" (METHOD 1)

pd.read_sql_query(
                    """
                        SELECT COUNT(*) AS Count_of_Vehicles_youngest_than_45 -- Count the number of vehicles younger than 45
                        FROM 
                            (
                            SELECT *,(2024-(1900+year)) AS vehicle_age        -- Calculate vehicle age
                            FROM results  
                            )
                        WHERE vehicle_age < 45                                -- Filter for vehicles younger than 45
                    """,connection)

Count_of_Vehicles_youngest_than_45
0                                  88

In [72]:
# 1.4 - How many vehicles are in our dataset that are under 45 and 45 and over? (METHOD 2)

pd.read_sql_query(
                    """
                        SELECT COUNT(*) AS Count_categorical_age,         -- Count the number of vehicles in each age category
                            CASE
                                WHEN vehicle_age >= 45 THEN '45+'         -- Assign '45+' if vehicle_age is 45 or more
                                ELSE '<45'                                -- Otherwise, assign '<45'
                            END Categorical_age                           -- Categorize the age into '45+' or '<45'
                        FROM 
                            (
                                SELECT *,(2024-(1900+year)) AS vehicle_age-- Calculate vehicle age
                                FROM results 
                             )
                        GROUP BY Categorical_age                          -- Group by the categorical age ('45+' and '<45')
                    """,connection)

Count_categorical_age Categorical_age
0                    309             45+
1                     88             <45

In [73]:
# 1.5 - All unique vehicle name that have 8 cylinders with horsepower that is greater than 125 but doesn't exceed 200

pd.read_sql_query(
                    """
                        SELECT DISTINCT name,cylinders,horsepower -- Select distinct combinations of name, cylinders, and horsepower
                        FROM results
                        WHERE cylinders = 8                       -- Where the number of cylinders is 8
                        AND horsepower BETWEEN 126 and 200        -- And the horsepower is between 126 and 200 (inclusive)
                    """,connection)

name  cylinders  horsepower
0              chevrolet chevelle malibu          8       130.0
1                      buick skylark 320          8       165.0
2                     plymouth satellite          8       150.0
3                          amc rebel sst          8       150.0
4                            ford torino          8       140.0
..                                   ...        ...         ...
75                 mercury grand marquis          8       138.0
76                       dodge st. regis          8       135.0
77               buick estate wagon (sw)          8       155.0
78              ford country squire (sw)          8       142.0
79  chrysler lebaron town @ country (sw)          8       150.0

[80 rows x 3 columns]

# Exercise 2 Songs Database

**2.1** Take the song_data.csv file and turn it into a single `sqlite` database. 

In [74]:
# Read a CSV file into a pandas DataFrame

df = pd.read_csv('../data/song_data.csv')

# Establish a connection to an SQLite database file

connection1 = sqlite3.connect('../Data/songs.db')

# Write the DataFrame to an SQLite database table

df.to_sql('songs_table',con = connection1,index = False,if_exists = 'replace')

# Select table names from the SQLite database

pd.read_sql("SELECT name FROM sqlite_master WHERE type = 'table'",connection1)

name
0  songs_table_updated
1          songs_table

In [75]:
# Select all data from the 'songs_table'

pd.read_sql_query('SELECT * FROM songs_table',connection1)

song_name  song_popularity  song_duration_ms  \
0      Boulevard of Broken Dreams               73            262333   
1                      In The End               66            216933   
2               Seven Nation Army               76            231733   
3                      By The Way               74            216933   
4               How You Remind Me               56            223826   
...                           ...              ...               ...   
18830              Let It Breathe               60            159645   
18831                     Answers               60            205666   
18832      Sudden Love (Acoustic)               23            182211   
18833           Gentle on My Mind               55            352280   
18834                    Up to Me               60            193533   

       acousticness  danceability  energy  instrumentalness  key  liveness  \
0          0.005520         0.496   0.682          0.000029    8    0.0589   
1          0.010300         0.542   0.853          0.000000    3    0.1080   
2          0.008170         0.737   0.463          0.447000    0    0.2550   
3          0.026400         0.451   0.970          0.003550    0    0.1020   
4          0.000954         0.447   0.766          0.000000   10    0.1130   
...             ...           ...     ...               ...  ...       ...   
18830      0.893000         0.500   0.151          0.000065   11    0.1110   
18831      0.765000         0.495   0.161          0.000001   11    0.1050   
18832      0.847000         0.719   0.325          0.000000    0    0.1250   
18833      0.945000         0.488   0.326          0.015700    3    0.1190   
18834      0.911000         0.640   0.381          0.000254    4    0.1040   

       loudness  audio_mode  speechiness    tempo  time_signature  \
0        -4.095           1       0.0294  167.060               4   
1        -6.407           0       0.0498  105.256               4   
2        -7.828           1       0.0792  123.881               4   
3        -4.938           1       0.1070  122.444               4   
4        -5.065           1       0.0313  172.011               4   
...         ...         ...          ...      ...             ...   
18830   -16.107           1       0.0348  113.969               4   
18831   -14.078           0       0.0301   94.286               4   
18832   -12.222           1       0.0355  130.534               4   
18833   -12.020           1       0.0328  106.063               4   
18834   -11.790           1       0.0302   91.490               4   

       audio_valence  
0              0.474  
1              0.370  
2              0.324  
3              0.198  
4              0.574  
...              ...  
18830          0.300  
18831          0.265  
18832          0.286  
18833          0.323  
18834          0.581  

[18835 rows x 15 columns]

# 2.2 Queries

**2.2.1** Use a single query to see if your favorite song is in this dataset

**2.2.2** How many unique popularity scores are there?

**2.2.3** How many song names are in this dataset that contain _'radio edit'_ in their name? Make sure your query is not case sensitive and there is only one single character (space) between _'radio'_ and _'edit'_.

**2.2.4** Write an SQL query to report songs with popularity of 99 and above. How many songs are there?

**2.2.5** Create a new column that contains song duration in the following format m:s. Example, the duration of Boulevard of Broken Dreams is 4:22

**2.2.6** Display the minimum, maximum and average popularity scores. The output of your query should contain only one row with 3 columns.

**2.2.7** Display average popularity scores and count for songs that contain "love" in their name and those that don't. Is there any difference in their popularity? 

**2.2.8** Display top 5 songs with the longest names.

In [76]:
# 2.2.1 Use a single query to see if your favorite song is in this dataset

# Define a variable with my favorite song's name

my_favorite_song = "Seven Nation Army"

# Select all data from the 'song_data' table where song_name matches my favorite song

pd.read_sql_query(f"""
                    SELECT *
                    FROM songs_table
                    WHERE song_name = "{my_favorite_song}"
                    """,connection1)

song_name  song_popularity  song_duration_ms  acousticness  \
0  Seven Nation Army               76            231733       0.00817   
1  Seven Nation Army               64            172919       0.89100   
2  Seven Nation Army               81            231920       0.00666   
3  Seven Nation Army               81            231920       0.00666   
4  Seven Nation Army               19            231800       0.00576   
5  Seven Nation Army               45            232626       0.00597   

   danceability  energy  instrumentalness  key  liveness  loudness  \
0         0.737   0.463          0.447000    0     0.255    -7.828   
1         0.690   0.283          0.000386    2     0.111    -7.309   
2         0.745   0.466          0.350000    4     0.272    -7.620   
3         0.745   0.466          0.350000    4     0.272    -7.620   
4         0.750   0.462          0.545000    0     0.307    -7.664   
5         0.745   0.464          0.410000    4     0.239    -7.587   

   audio_mode  speechiness    tempo  time_signature  audio_valence  
0           1       0.0792  123.881               4          0.324  
1           1       0.0294  104.348               4          0.350  
2           0       0.0864  123.889               4          0.303  
3           0       0.0864  123.889               4          0.303  
4           1       0.0779  123.865               4          0.309  
5           0       0.0820  123.893               4          0.297

In [77]:
# 2.2.2 How many unique popularity scores are there?

pd.read_sql_query(
                    """
                        SELECT COUNT(DISTINCT song_popularity) AS Count_of_unique_popularity_scores -- Count the number of unique song popularity scores
                        FROM songs_table
                    """,connection1)

Count_of_unique_popularity_scores
0                                101

In [78]:
# 2.2.3 How many song names are in this dataset that contain 'radio edit' in their name?

pd.read_sql_query(
                    """
                        SELECT COUNT(song_name) AS 'Count radio edit' -- Count the number of occurrences of song names containing 'radio edit'
                        FROM songs_table
                        WHERE song_name LIKE '%radio edit%'           -- Where the song name contains 'radio edit'
                    """,connection1)

Count radio edit
0               132

In [79]:
# 2.2.4 - Write an SQL query to report songs with popularity of 99 and above. 

pd.read_sql_query(
                    """
                        SELECT  DISTINCT song_name,song_popularity -- Select distinct combinations of song_name and song_popularity
                        FROM songs_table
                        WHERE song_popularity >= 99                -- Where the song popularity is 99 or higher
                    """,connection1)

song_name  song_popularity
0  I Love It (& Lil Pump)               99
1                 Happier              100

In [80]:
# 2.2.4 - How many songs are there?

pd.read_sql_query(
                    """
                        SELECT COUNT (*) AS songs_score_above_99          -- Count the number of distinct songs with popularity 99 or higher
                        FROM
                            (
                                SELECT DISTINCT song_name,song_popularity -- Select distinct combinations of song_name and song_popularity
                                FROM songs_table
                                WHERE song_popularity >= 99               -- Where the song popularity is 99 or higher
                            )""",connection1)

songs_score_above_99
0                     2

In [81]:
# 2.2.5 - Create a new column that contains song duration in the following format m:s. Example, the duration of Boulevard of Broken Dreams is 4:22

df = pd.read_sql_query('SELECT * FROM songs_table', connection1)

# Convert song duration from milliseconds to seconds

df['duration_sec'] = df['song_duration_ms'] / 1000

# Convert seconds to m:s format

df['duration_m_s'] = pd.to_datetime(df['duration_sec'], unit='s').dt.strftime('%M:%S')

# Update SQLite table with the new column

df.to_sql('songs_table_updated', con=connection1, index=False, if_exists='replace')

# Select data from 'songs_table_updated' to verify the new column is present

pd.read_sql_query('SELECT song_name, duration_m_s FROM songs_table_updated', connection1)

song_name duration_m_s
0      Boulevard of Broken Dreams        04:22
1                      In The End        03:36
2               Seven Nation Army        03:51
3                      By The Way        03:36
4               How You Remind Me        03:43
...                           ...          ...
18830              Let It Breathe        02:39
18831                     Answers        03:25
18832      Sudden Love (Acoustic)        03:02
18833           Gentle on My Mind        05:52
18834                    Up to Me        03:13

[18835 rows x 2 columns]

In [82]:
# 2.2.6 The minimum, maximum, and average popularity scores

pd.read_sql_query(
                    """
                        SELECT MIN(song_popularity) AS min_popularity_scores, -- Select the minimum song popularity score
                        MAX (song_popularity) AS max_popularity_scores,       -- Select the maximum song popularity score
                        AVG(song_popularity) AS avg_popularity_scores         -- Select the average song popularity score
                        FROM songs_table
                    """,connection1)

min_popularity_scores  max_popularity_scores  avg_popularity_scores
0                      0                    100              52.991877

In [83]:
# 2.2.7 - The average popularity scores and count for songs that contain "love" 

pd.read_sql_query(
                    """
                        SELECT AVG(song_popularity) AS avg_love_songs      -- Select the average song popularity
                        FROM
                            (
                                SELECT DISTINCT song_name, song_popularity -- Select distinct combinations of song_name and song_popularity
                                FROM songs_table
                                WHERE song_name LIKE '%love%'              -- Where the song name contains 'love'  
                            )
                    """,connection1)

avg_love_songs
0       49.025271

In [84]:
# 2.2.7 - Average popularity scores and count for songs that do not contain 'love' in their name

pd.read_sql_query(
                    """
                        SELECT AVG(song_popularity) AS avg_no_love_songs      -- Select the average song popularity
                        FROM
                            (
                                SELECT DISTINCT song_name, song_popularity -- Select distinct combinations of song_name and song_popularity
                                FROM songs_table
                                WHERE song_name <> 'love'                  --  Where the song name is not equal to 'love'
                            )
                    """,connection1)

avg_no_love_songs
0          48.734457

In [85]:
# 2.2.7 - Is there any difference in their popularity?

avg_love_songs - avg_no_love_songs = 49.025271 - 48.734457

                                = 0.29081399999999746
There is almost no difference. We could say it's negligible.

In [86]:
# 2.2.8 Top 5 songs with the longest names

pd.read_sql_query(
                    """
                        SELECT DISTINCT song_name,LENGTH("song_name") AS Length_name_song -- Select distinct song names and their lengths
                        FROM songs_table
                        ORDER BY Length_name_song DESC                                    -- Order the results by song name length in descending order
                        LIMIT 5                                                           -- Limit the results to the top 5 longest song names
                    """,connection1)

song_name  Length_name_song
0  Viajo Sin Ver (Remix) [feat. De La Ghetto, Alm...               143
1  Riff Off: Mickey/Like A Virgin/Hit Me With You...               141
2  Wu Tang Forever (ft. Ghostface Killah, Raekwon...               139
3  Bellas Finals: Price Tag/Don’t You (Forget Abo...               133
4  I Don’t Wanna Live Forever (Fifty Shades Darke...               114